File: EmployeeImportance(BFS).java
//Time Complexity : O(n), All the Employees
//Space Complexity : O(n),All the Employees 
//Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : no


/*
// Definition for Employee.
class Employee {
    public int id;
    public int importance;
    public List<Integer> subordinates;
};
*/

class Solution {
    HashMap<Integer, Employee> map;
    int totalImp;
    public int getImportance(List<Employee> employees, int id) {
        if(employees == null)return 0;
        map = new HashMap<>();
        totalImp = 0;
        Queue<Integer> q = new LinkedList<>();
        for(Employee e : employees){
            map.put(e.id,e);
        }
        q.add(id);
        while(!q.isEmpty()){
            int currEid = q.poll();
            Employee currEmp = map.get(currEid);
            totalImp += currEmp.importance;
            for(int subs : currEmp.subordinates ){
                q.add(subs);
            }
            
        }
        return totalImp;
    }
}

File: Employee_Importance(BFS).cpp
Time Complexity-O(n)
Space Complexity-O(n)
Did the code run on Leetcode? yes

class Solution {
public:
    int getImportance(vector<Employee*> employees, int id) {
        int res=0;
        unordered_map<int,Employee*>m;
        for(int i=0;i<employees.size();i++)
        {
            m[employees[i]->id]=employees[i];
        }
        queue<int>k;
        k.push(id);
        while(!k.empty())
        {
            int size=k.size();
            while(size!=0)
            {
                int temp=k.front();
                k.pop();
                if(m.find(temp)!=m.end())
                {
                    res+=m[temp]->importance;
                    for(int i=0;i<m[temp]->subordinates.size();i++)
                    {
                        k.push(m[temp]->subordinates[i]);
                    }
                }
                size--;
            }
        }
        return res;
    }
};
